core/trino-parser/src/test/java/io/trino/sql/parser/ParserAssert.java

```diff
@@ -99,6 +99,11 @@ public ParserAssert ignoringLocation()
         return ignoringFieldsMatchingRegexes("(.*\\.)?location");
     }
 
+    public ParserAssert ignoringHash()
```

Comment:
Why do we need this?

Reply:
Without it tests in TestTypeSignatureTranslator fails. Hashes are different if case of types is different, e.g:
```
java.lang.AssertionError: 
Expecting:
  <ROW(a BIGINT, b VARCHAR)>
to be equal to:
  <ROW(a bigint, b varchar)>
when recursively comparing field by field, but found the following 3 differences:

field/property 'fields.type.hash' differ:
- actual value   : -472292219
- expected value : -1254919067
```

Reply:
That's just masking an underlying problem. I don't think we should have a method for ignoring the hash in tests. It's actually an implementation detail whether the hash field exists or not in the class, so with this, we're 1) coupling tests to physical structure of those classes, 2) potentially masking bugs, 3) asking test writers to think about hashes should be equal or not.

Reply:
So what would you propose?
Also, I'm not sure what do you mean by "underlaying problem":
- We normally do not check hash when testing equality of expressions. 
- 'a BIGINT' is not equal to 'a bigint' when using equals() method
-  The comparison in TestTypeSignatureTranslator is artificial, it directly accesses all fields of objects and compare them. We already ignore 'location' attribute, why can't we also ignore 'hash'?

Reply:
We ignore location because the syntax tree currently serves two mixed purposes: it's the representation for the syntax tree, which includes locations, and it's also the representation for the IR (intermediate representation), which does not include locations. So depending on which world the code lives in, testing for locations is appropriate (or not)

Reply:
> So depending on which world the code lives in, testing for locations is appropriate (or not)

Does it mean that we should have dedicated `equals` methods defined in `Node`?

Reply:
No, because the analyzer/AST doesn't care about pure equality (aside than for tests to ensure the parser is doing the right thing). The analyzer uses CanonicalizationAware and ScopeAware to do semantically relevant comparisons between trees.